std/socket.d

```diff
@@ -307,56 +307,94 @@ bool wouldHaveBlocked() nothrow @nogc
         static assert(0);
 }
 
+// Returns true iff getnameinfo is supported on this platform.
+version(Windows) private @property bool getnameinfoSupported()
+{
+    initialize;
+    return getNameInfoPointer !is null;
+}
+// Ditto (always supported on Posix).
+version(Posix) private enum bool getnameinfoSupported = true;
 
-private immutable
+// Returns true iff getaddrinfo is supported on this platform.
+version(Windows) private @property bool getaddrinfoSupported()
 {
-    typeof(&getnameinfo) getnameinfoPointer;
-    typeof(&getaddrinfo) getaddrinfoPointer;
-    typeof(&freeaddrinfo) freeaddrinfoPointer;
+    initialize;
+    return getaddrInfoPointer !is null;
 }
+// Ditto (always supported on Posix).
+version(Posix) private enum bool getaddrinfoSupported = true;
 
-shared static this() @system
+version(Windows) private
 {
-    version(Windows)
-    {
-        WSADATA wd;
+    import core.atomic;
 
-        // Winsock will still load if an older version is present.
-        // The version is just a request.
-        int val;
-        val = WSAStartup(0x2020, &wd);
-        if (val)         // Request Winsock 2.2 for IPv6.
-            throw new SocketOSException("Unable to initialize socket library", val);
+    shared typeof(&getnameinfo) getnameinfoPointer;
+    shared typeof(&getaddrinfo) getaddrinfoPointer;
+    shared typeof(&freeaddrinfo) freeaddrinfoPointer;
 
-        // These functions may not be present on older Windows versions.
-        // See the comment in InternetAddress.toHostNameString() for details.
-        auto ws2Lib = GetModuleHandleA("ws2_32.dll");
-        if (ws2Lib)
-        {
-            getnameinfoPointer = cast(typeof(getnameinfoPointer))
-                                 GetProcAddress(ws2Lib, "getnameinfo");
-            getaddrinfoPointer = cast(typeof(getaddrinfoPointer))
-                                 GetProcAddress(ws2Lib, "getaddrinfo");
-            freeaddrinfoPointer = cast(typeof(freeaddrinfoPointer))
-                                 GetProcAddress(ws2Lib, "freeaddrinfo");
-        }
+    auto our_getnameinfo(A...)(A a) @system
+    {
+        initialize;
+        return atomicLoad(getnameinfoPointer)(a);
     }
-    else version(Posix)
+
+    auto our_getaddrinfo(A...)(A a) @system
     {
-        getnameinfoPointer = &getnameinfo;
-        getaddrinfoPointer = &getaddrinfo;
-        freeaddrinfoPointer = &freeaddrinfo;
+        initialize;
+        return atomicLoad(getaddrinfoPointer)(a);
     }
-}
 
+    auto our_freeaddrinfo(A...)(A a) @system
+    {
+        initialize;
+        return atomicLoad(freeaddrinfoPointer)(a);
+    }
 
-shared static ~this() @system nothrow @nogc
-{
-    version(Windows)
+    private void initialize() @trusted
     {
-        WSACleanup();
+        static bool tlsInitialized;
+        if (tlsInitialized) return;
```

Comment:
???
Why do you put a *TLS* flag around the double-checked locking in initOnce?
https://github.com/dlang/phobos/blob/27072f8282fd60ef2af025280d217729d2b5770f/std/concurrency.d#L2487
Considering how slow name lookups are, performance shouldn't be a hindrance. And even then you'd better spent some time to optimize initOnce for dmd than to add more and more workarounds.

Reply:
Just wanted to make `initialize` cheap to call in all instances.

Reply:
It's already cheap to call initOnce, that's the purpose of it.
And besides a TLS access isn't actually free either.

Reply:
I'm a bit worried about cargo cult here.
That TLS locking pattern came up as alternative to "double-checked locking", only that double-checked locking works fine with atomic acquire&release primitives (which are free on X86).
So in total you have the overhead of a TLS access against a normal X86 load instruction.
Now in this PR you added both to be even faster.?

Reply:
No problem, glad to remove.